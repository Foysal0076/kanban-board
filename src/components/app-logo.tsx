import { AppIcon } from '@/icons'

const AppLogo = () => {
  return (
    <div className='flex items-center justify-center gap-2 tracking-tighter text-primary'>
      <AppIcon className='h-6 w-6 md:h-[1.625rem] md:w-[1.625rem]' />
      <h1 className='text-[1.75rem] font-semibold leading-none text-foreground md:text-[2.125rem]'>
        kanban
      </h1>
    </div>
  )
}

export default AppLogo
