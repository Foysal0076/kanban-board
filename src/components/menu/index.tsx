'use client'
import { usePathname } from 'next/navigation'
import * as React from 'react'
import {
  useCallback,
  useContext,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
} from 'react'

import useClickOutside from '@/hooks/use-click-outside'

// 1. Popover - holds the state and methods, and expose it through context
// 2. Trigger - attach the trigger method to children
// 3. Content - render conditionally based on the state, through context
// 4. Close - attach the close method to children

// TODO: cover more positions
type Position = 'left' | 'right' | 'center'

const defaultRect: DOMRect = {
  left: 0,
  top: 0,
  width: 0,
  height: 0,
  x: 0,
  y: 0,
  bottom: 0,
  right: 0,
  toJSON: () => ({}),
}

const MenuContext = React.createContext<{
  isShow: boolean
  setIsShow: React.Dispatch<React.SetStateAction<boolean>>
  preferredPosition: Position
  triggerRect: DOMRect
  setTriggerRect: React.Dispatch<React.SetStateAction<DOMRect>>
}>({
  isShow: false,
  setIsShow: () => {
    throw new Error('MenuContext setIsShow should be used under provider')
  },
  preferredPosition: 'right',
  triggerRect: defaultRect,
  setTriggerRect: () => {
    throw new Error('MenuContext setTriggerRect should be used under provider')
  },
})

export default function Menu({
  children,
  preferredPosition = 'right',
}: {
  children: React.ReactNode
  preferredPosition?: Position
}) {
  const [isShow, setIsShow] = useState(false)
  const [triggerRect, setTriggerRect] = useState(defaultRect)

  const pathname = usePathname()

  const contextValue = {
    isShow,
    setIsShow,
    preferredPosition,
    triggerRect,
    setTriggerRect,
  }

  useEffect(() => {
    setIsShow(false)
  }, [pathname])

  return (
    <MenuContext.Provider value={contextValue}>{children}</MenuContext.Provider>
  )
}

function Trigger({ children }: { children: React.ReactElement }) {
  const { setIsShow, setTriggerRect } = useContext(MenuContext)

  const ref = useRef<HTMLElement>(null)

  const onClick = (_e: MouseEvent) => {
    const element = ref.current
    if (element == null) {
      return
    }

    const rect = element.getBoundingClientRect()
    setTriggerRect(rect)
    setIsShow((isShow) => !isShow)
  }

  const childrenToTriggerPopover = React.cloneElement(children, {
    onClick, // TODO: we better merge the onClick
    ref, // TODO: we better merge the ref
  })

  return childrenToTriggerPopover
}

function Content({ children }: { children: React.ReactNode }) {
  const { isShow } = useContext(MenuContext)

  if (!isShow) {
    return null
  }

  return <ContentInternal>{children}</ContentInternal>
}

function ContentInternal({ children }: { children: React.ReactNode }) {
  const { triggerRect, preferredPosition, setIsShow } = useContext(MenuContext)
  const ref = useRef<HTMLDialogElement>(null)
  const [coords, setCoords] = useState({
    left: 0,
    top: 0,
    transform: 'none',
  })

  useLayoutEffect(() => {
    const element = ref.current
    if (element == null) {
      return
    }

    const rect = element.getBoundingClientRect()

    const coords = getPopoverCoords(triggerRect, rect, preferredPosition)
    setCoords(coords)
  }, [])

  const refFocusTrapping = useFocusTrapping()

  const dismiss = useCallback(() => {
    setIsShow(false)
  }, [])
  const refClickOutside = useClickOutside(ref, dismiss)

  const mergedRef = mergeRef(ref, refFocusTrapping, refClickOutside)

  return (
    <dialog
      open={true}
      ref={mergedRef}
      style={{
        position: 'fixed',
        zIndex: 9999,
        left: `${coords.left}px`,
        top: `${coords.top}px`,
        transform: coords.transform,
        margin: 0,
      }}>
      {children}
    </dialog>
  )
}

function Close({ children }: { children: React.ReactElement }) {
  const { setIsShow } = useContext(MenuContext)
  const onClick = (e: MouseEvent) => {
    setIsShow(false)

    // popover will be gone
    // prevent this event triggering unexpected click
    e.stopPropagation()
  }
  const childrenToClosePopover = React.cloneElement(children, {
    onClick, // TODO: we better merge the onClick
  })

  return childrenToClosePopover
}

Menu.Trigger = Trigger
Menu.Content = Content
Menu.Close = Close

function getPopoverCoords(
  triggerRect: DOMRect,
  popoverRect: DOMRect,
  position: Position
) {
  const offset = 4
  let top: number = triggerRect.top + triggerRect.height + offset
  let left: number
  let transform: string = 'none'

  switch (position) {
    case 'left':
      left = triggerRect.left - offset - popoverRect.width
      break
    case 'center':
      left = triggerRect.left + triggerRect.width / 2 - popoverRect.width / 2
      break
    case 'right':
    default:
      left = triggerRect.right + offset
      break
  }

  const overflowAtLeft = left < offset
  const overflowAtRight = left + popoverRect.width > window.innerWidth
  const overflowAtTop = top < offset
  const overflowAtBottom = top + popoverRect.height > window.innerHeight

  // If content overflow at the bottom, show on top
  if (overflowAtBottom) {
    top = triggerRect.top - offset - popoverRect.height
  }

  // If content overflow at the right, show on left
  if (overflowAtRight) {
    left = triggerRect.left - offset - popoverRect.width
  }

  // If content overflow at the left, show on right
  if (overflowAtLeft) {
    left = triggerRect.right + offset
  }

  // If content overflow at the top, show on bottom
  if (overflowAtTop) {
    left = triggerRect.left - offset
  }

  return {
    top,
    left,
    transform,
  }
}

// TODO: better focusable query
const focusableQuery = ':is(input, button, [tab-index]'

// some hooks
function useFocusTrapping() {
  // @ts-ignore TODO: fix the typings
  const refTrigger = useRef<HTMLElement>(document.activeElement)
  const ref = useRef<HTMLElement>(null)

  const onKeyDown = useCallback((e: KeyboardEvent) => {
    const popover = ref.current
    if (popover == null) {
      return
    }
    const focusables = [...popover.querySelectorAll(focusableQuery)]

    switch (e.key) {
      case 'Tab':
        // check if it is the last focusable
        const lastFocusable = focusables[focusables.length - 1]
        if (document.activeElement === lastFocusable) {
          // @ts-ignore, TODO: fix typing
          focusables[0]?.focus()

          e.preventDefault()
        }
    }
  }, [])

  useEffect(() => {
    const popover = ref.current
    if (popover == null) {
      return
    }

    const focusables = [...popover.querySelectorAll(focusableQuery)]
    // 1. focus the first focusable
    // @ts-ignore, TODO: fix typing
    focusables[0]?.focus()
    // console.log('mount popover focusing', focusables[0])

    // 2. attach keyboard event listener to trap the focus
    document.addEventListener('keydown', onKeyDown)
    return () => {
      document.removeEventListener('keydown', onKeyDown)

      // 3. refocus the trigger after dismissing
      // but only if the current activeElement is body
      // since this happens after popover is gone
      // am I right about this?
      const trigger = refTrigger.current
      const currentActiveElement = document.activeElement
      if (currentActiveElement == document.body) {
        trigger?.focus()
      }
    }
  }, [])

  return ref
}

function mergeRef<T>(
  ...refs: Array<React.MutableRefObject<T> | React.RefCallback<T>>
) {
  return (el: T) => {
    refs.forEach((ref) => {
      if (typeof ref === 'function') {
        ref(el)
      } else {
        ref.current = el
      }
    })
  }
}
