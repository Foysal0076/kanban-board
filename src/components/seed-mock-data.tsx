'use client'

import { useEffect } from 'react'

import { STORAGE_KEYS } from '@/config/constants'
import { MOCK_BOARDS, MOCK_USERS } from '@/config/mock-data'
import { getAllBoards } from '@/features/board/service/board.service'
import { User } from '@/types/session-user.type'
import { getLocalStorage, setLocalStorage } from '@/utils'

export const SeedMockData = () => {
  useEffect(() => {
    //check if users already exist in localStorage
    const users = getLocalStorage<User[]>(STORAGE_KEYS.USERS)
    if (!users || users.length === 0) {
      setLocalStorage(STORAGE_KEYS.USERS, MOCK_USERS)
    }

    const boards = getAllBoards()
    if (!boards || boards.length === 0) {
      setLocalStorage(STORAGE_KEYS.BOARDS, MOCK_BOARDS)
    }
  }, [])

  return null
}
