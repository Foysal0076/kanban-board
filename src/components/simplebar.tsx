'use client'

import 'simplebar-react/dist/simplebar.min.css'

import SimpleBar from 'simplebar-react'

export default SimpleBar
