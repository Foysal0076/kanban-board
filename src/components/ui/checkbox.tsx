import { forwardRef, InputHTMLAttributes, useId } from 'react'

import { cn } from '@/utils'

type RadioProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'type'> & {
  label?: string
  containerClassName?: string
}

export const Checkbox = forwardRef<HTMLInputElement, RadioProps>(
  ({ className, label, containerClassName = '', id, ...props }, ref) => {
    const _id = useId()

    return (
      <label
        htmlFor={`${id}_${_id}`}
        className={cn(
          'flex cursor-pointer items-center space-x-2',
          containerClassName
        )}>
        <input
          {...props}
          ref={ref}
          type='checkbox'
          id={`${id}_${_id}`}
          className={cn(
            'h-4 w-4 border-border checked:bg-primary focus:ring-2 focus:ring-primary',
            className
          )}
        />
        <span className='select-none text-sm text-foreground'>{label}</span>
      </label>
    )
  }
)
