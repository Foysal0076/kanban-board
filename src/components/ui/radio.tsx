import { forwardRef, InputHTMLAttributes } from 'react'

import { cn } from '@/utils'

type RadioProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'type'> & {
  label?: string
}

export const Radio = forwardRef<HTMLInputElement, RadioProps>(
  ({ className, label, id, ...props }, ref) => {
    return (
      <div className='flex items-center space-x-2'>
        <input
          {...props}
          ref={ref}
          type='radio'
          id={id}
          className={cn(
            'text-primary-600 h-4 w-4 border-border bg-accent focus:ring-2 focus:ring-primary',
            className
          )}
        />
        {label && (
          <label
            htmlFor={id}
            className='cursor-pointer select-none text-sm text-foreground'>
            {label}
          </label>
        )}
      </div>
    )
  }
)
