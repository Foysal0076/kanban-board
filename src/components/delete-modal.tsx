'use client'

import { useModal } from '@/components/modal/use-modal'
import { Button } from '@/components/ui'
import { TrashIcon } from '@/icons'

type Props = {
  onConfirmDelete: () => void
  onClose?: () => void
  titleText?: string
  bodyText?: string
  cancelButtonText?: string
  deleteButtonText?: string
}

export default function DeleteModal({
  bodyText = 'Are you sure you want to archive this item? You can always unarchive later',
  onClose,
  onConfirmDelete,
  titleText = 'Archive this item?',
  cancelButtonText = 'Cancel',
  deleteButtonText = 'Delete',
}: Props) {
  const { closeModal } = useModal()
  const _onClose = onClose || closeModal

  return (
    <div className='card flex flex-col items-center gap-2.5 p-4 max-sm:w-[90vw] sm:w-[24rem]'>
      <TrashIcon className='h-10 w-10 text-destructive' />
      <h2 className='font-bold'>{titleText}</h2>
      <p className='text-center text-muted-foreground'>{bodyText}</p>
      <div className='mt-2 flex w-full gap-2'>
        <Button className='w-1/2' variant={'outline'} onClick={_onClose}>
          {cancelButtonText}
        </Button>
        <Button
          className='w-1/2'
          variant={'destructive'}
          onClick={onConfirmDelete}>
          {deleteButtonText}
        </Button>
      </div>
    </div>
  )
}
