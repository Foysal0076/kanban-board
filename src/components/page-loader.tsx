import { Spinner } from '@/components/ui'

export default function PageLoader() {
  return (
    <div className='flex min-h-[70vh] items-center justify-center'>
      <Spinner className='text-primary' />
    </div>
  )
}
