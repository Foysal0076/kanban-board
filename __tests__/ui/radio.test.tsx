import { fireEvent, render, screen } from '@testing-library/react'

import { Radio } from '@/components/ui'

describe('Radio', () => {
  const onChange = jest.fn()

  it('renders correctly with default props', () => {
    render(<Radio onChange={onChange} />)
    const radio = screen.getByRole('radio')
    expect(radio).toBeInTheDocument()
  })

  it('applies custom className correctly', () => {
    render(<Radio className='custom-class' onChange={onChange} />)
    const radio = screen.getByRole('radio')
    expect(radio).toHaveClass('custom-class')
  })

  it('forwards ref correctly', () => {
    const ref = jest.fn()
    render(<Radio ref={ref} onChange={onChange} />)
    expect(ref).toHaveBeenCalled()
  })

  it('handles disabled state', () => {
    render(<Radio disabled onChange={onChange} />)
    const radio = screen.getByRole('radio')
    expect(radio).toBeDisabled()
  })

  it('handles user input correctly', () => {
    render(<Radio onChange={onChange} />)
    const radio = screen.getByRole('radio')
    fireEvent.click(radio)
    expect(onChange).toHaveBeenCalled()
    expect(radio).toBeChecked()
  })

  it('renders label correctly', () => {
    render(<Radio label='Test Label' onChange={onChange} />)
    const label = screen.getByText('Test Label')
    expect(label).toBeInTheDocument()
  })

  // Snapshot test
  it('renders correctly with default props', () => {
    const { container } = render(<Radio onChange={onChange} />)
    expect(container).toMatchSnapshot()
  })
})
