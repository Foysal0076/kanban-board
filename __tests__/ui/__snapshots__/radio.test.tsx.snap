// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders correctly with default props 1`] = `
<div>
  <div
    class="flex items-center space-x-2"
  >
    <input
      class="text-primary-600 h-4 w-4 border-border bg-accent focus:ring-2 focus:ring-primary"
      type="radio"
    />
  </div>
</div>
`;
